### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by riusricardo**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -2,9 +2,9 @@ resolver = "2"
 
 [workspace]
 members = [
-    'node',
-    'runtime',
+    "node",
```
Comment (HCastano) [inline Cargo.toml:5]: nit: Single quotes were being used in all the TOML files since that kept the diff between `canvas` and `substrate-parachain-template` (our reference) small. 

For that reason I'd prefer to keep the single quotes, but would be open to changing them if you have a good reason as to why. Either way, this should be done in a standalone PR since it makes the diff pretty noisy

File: runtime/Cargo.toml [hunk 2]
```diff
@@ -136,14 +138,17 @@ contracts-unstable-interface = [
...
-	'frame-system-benchmarking',
-	'frame-support/runtime-benchmarks',
-	'frame-system/runtime-benchmarks',
-	'pallet-balances/runtime-benchmarks',
-	'pallet-timestamp/runtime-benchmarks',
-	'pallet-xcm/runtime-benchmarks',
-	'pallet-collator-selection/runtime-benchmarks',
+	"hex-literal",
+	"sp-runtime/runtime-benchmarks",
+	"xcm-builder/runtime-benchmarks",
```
Comment (HCastano) [inline runtime/Cargo.toml:143]: nit: This is easier to skim sorted 😉

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -858,8 +858,10 @@ impl pallet_contracts_rpc_runtime_api::ContractsApi<Block, AccountId, Balance, B
 
 			list_benchmark!(list, extra, frame_system, SystemBench::<Runtime>);
 			list_benchmark!(list, extra, pallet_balances, Balances);
+			list_benchmark!(list, extra, pallet_session, SessionBench::<Runtime>);
 			list_benchmark!(list, extra, pallet_timestamp, Timestamp);
 			list_benchmark!(list, extra, pallet_collator_selection, CollatorSelection);
+			list_benchmark!(list, extra, pallet_contracts, Contracts);
```
Comment (HCastano) [inline runtime/src/lib.rs:864]: I'm a little OOTL here, what are metadata benchmarks?

File: node/src/command.rs [hunk 4]
```diff
@@ -39,9 +39,9 @@ fn load_spec(
 	para_id: ParaId,
 ) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
 	Ok(match id {
-		"dev" => Box::new(chain_spec::development_config(para_id)),
+		"canvas-dev" => Box::new(chain_spec::development_config(para_id)),
```
Comment (HCastano) [inline node/src/command.rs:42]:
```suggestion
		"dev | canvas-dev" => Box::new(chain_spec::development_config(para_id)),
```

I don't want to lose the ability to run with `--dev` (which I believe is checked here)

File: node/src/command.rs [hunk 5]
```diff
@@ -39,9 +39,9 @@ fn load_spec(
...
 		"canvas-rococo" => Box::new(chain_spec::rococo_testnet_config(para_id)),
-		"" | "local" => Box::new(chain_spec::local_testnet_config(para_id)),
+		"canvas-local" => Box::new(chain_spec::local_testnet_config(para_id)),
```
Comment (HCastano) [inline node/src/command.rs:44]:
```suggestion
		"local | canvas-local" => Box::new(chain_spec::local_testnet_config(para_id)),
```

Same thing here

File: runtime/Cargo.toml [hunk 6]
```diff
@@ -102,14 +104,14 @@ std = [
...
 	"pallet-aura/std",
-	"pallet-sudo/std",
 	"pallet-balances/std",
 	"pallet-collator-selection/std",
 	"pallet-contracts-primitives/std",
 	"pallet-contracts-rpc-runtime-api/std",
 	"pallet-contracts/std",
 	"pallet-randomness-collective-flip/std",
 	"pallet-session/std",
+	"pallet-sudo/std",
```
Comment (HCastano) [inline runtime/Cargo.toml:114]: I learned last week that we can't compile without this even though we don't use `pallet-sudo` ourselves (something about the `polkadot-test-runtime`), do you know anything more about why that is?

File: scripts/benchmarks.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
```
Comment (HCastano) [inline scripts/benchmarks.sh:2]:
```suggestion

set -eux

```

File: scripts/benchmarks.sh [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+steps=50
```
Comment (HCastano) [inline scripts/benchmarks.sh:3]: Since all these variables are only used once I'd just hardcode them into the bench command

File: scripts/benchmarks.sh [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	pallet_balances
+	pallet_collator_selection
+	pallet_contracts
+	pallet_session
+	pallet_timestamp
+)
+
+for p in ${pallets[@]}
+do
+	./target/release/canvas benchmark \
```
Comment (HCastano) [inline scripts/benchmarks.sh:17]:
```suggestion
	cargo run --release --features=runtime-benchmarks -- benchmark \
```

This is probably okay with the caching between runs, but if not the script should at least build the binary beforehand instead of making the caller of the script do that themselves

File: scripts/benchmarks.sh [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+canvasChain=canvas-dev
+pallets=(
+	pallet_balances
+	pallet_collator_selection
+	pallet_contracts
+	pallet_session
+	pallet_timestamp
+)
+
+for p in ${pallets[@]}
```
Comment (HCastano) [inline scripts/benchmarks.sh:15]:
```suggestion
for pallet in ${pallets[@]}
```

File: scripts/benchmarks.sh [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+	./target/release/canvas benchmark \
+		--chain=$canvasChain \
+		--execution=wasm \
+		--wasm-execution=compiled \
+		--pallet=$p  \
+		--extrinsic='*' \
+		--steps=$steps  \
+		--repeat=$repeat \
+		--raw  \
+		--output=$canvasOutput
```
Comment (HCastano) [inline scripts/benchmarks.sh:26]: Don't we also need a `template` for the Rust weight output files?

File: scripts/benchmarks.sh [hunk 12]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (HCastano) [inline scripts/benchmarks.sh:1]: I'd add some usage instructions for the script

**Comment by HCastano** [comment]
Oh, can you also add a CI stage which `check`'s the build with `--features runtime-benchmarks`?

**Comment by cmichi** [comment]
What is the status of this PR?

Side-note: With my recent `cumulus` sync in https://github.com/paritytech/canvas/pull/122 I left out a commit which touched benchmarks in `cumulus/parachain-template`: https://github.com/paritytech/cumulus/commit/2327eae5c9ea0724a033f73707b9135bcac21ab4#diff-ec2df2e97e27873c3143845320b3b5175fcf92bc76a103f4a1755745cb3cf049 (because the `master` currently has no benchmarks). So the changes from this commit should also be integrated into this PR.

**Comment by riusricardo** [comment]
This PR is very outdated now. It was blocked by a bug in the Collator-selection pallet which had broken benchmarks. It is fixed now and we should be good to add benchmarks.

**Comment by athei** [comment]
Runtime has moved to the cumulus repo. I started from scratch using the statemine runtime there. I think the benchmarking infrastructure is already in place now.
